# Анализ вакансий DS, DA в телеграм каналах

#  План работы над проектом

**Цель проекта**:
* Составить портрет сотрудника, которого чаще всего готовы принять на позиции DS, DA, а также оценить условия труда, предлогаемые в этих вакансиях. На основе анализа можно будет проследить как меняется рынок труда, выделить тенденции. Это поможет соискателям лучше понимать нанимающие компании, оценить и сопоставить свои ожидания и предложения компаний.

**При описании портрета сотрудника и оценке условий труда, будем учитывать следующее:**
* образование
* нанимающую компанию
* размер компании
* опыт сотрудника (грейд)
* требуемый стек технологий
* формат работы (удаленный/офис)
* зарплата

**В ходе исследования проверим следующие гипотезы:**
1. Больше всего вакансий, предлагающих работу в офисе, чем удаленный формат.
2. Приблизиттельное одинаковое количество вакансий для DA и DS
3. Больше вакансий от крупных компаний, чем от стартапов. 
4. Больше всего вакансий для специалистов с опытом, чем для junior специалистов.
5. В зимнее месяцы количество вакансий больше, чем в летние месяцы. 




# Этапы выполнения проекта.
1. Парсер телеграм-каналов с вакансиями DS, DA.
        * выбрать необходимые телеграм-каналы
        * выбрать период за который будут анализироваться вакансии
        * установить лимит выгружаемых вакансий
        * сформировать csv-файл с первоначальными данными о вакансиях
2. Предобработка данных:
        * Очистить и подготовить текст для дальнейшей работы (с использованием регулярных выражений, лемматизация, библиотеки    ntlk с стоп-словами).
        * Исследовать пропущенные значения
        * Проверить наличие дубликатов
        * Работа с типами данных
        * Создание необходимых колонок
3. Исследовательский анализ данных:
        * Обнаружение аномалий, выбросов, интересных фактов и зависимостей.
        * Поиск ответов на выдвинутые гипотезы
        * Визуальзация данных с 
4. Оформление результатов исследования в понятный и функциональный дашборд. 
        * визуализировать ответы на выдвинутые гипотезы
        * визуализировать потррет типичного специалиста DS и DA

Для анализа вакансий будем использовать следующие чаты.

* https://t.me/foranalysts
* https://t.me/analyst_job
* https://t.me/biheadhunter
* https://t.me/datasciencejobs
* https://t.me/bds_job
* https://t.me/datajobschannel
* https://t.me/data_hr

# Парсер телеграм-каналов

In [1]:
# !pip3 install pyrogram tgcrypto

In [2]:
# !pip3 install python-dotenv

In [3]:
pip install emoji

Note: you may need to restart the kernel to use updated packages.


In [4]:
import pyrogram
import requests
import numpy as np
from bs4 import BeautifulSoup
from dotenv import load_dotenv
import os
from pyrogram import Client
import pandas as pd
import json
from datetime import datetime
import re
from IPython.core.display import display, HTML, clear_output
import ipywidgets as widgets
from urllib.parse import urlparse
import emoji

/var/folders/vf/z1gk_tyn2zsf73w6bspmxgjm0000gn/T/ipykernel_81140/844583093.py:12: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML, clear_output
/var/folders/vf/z1gk_tyn2zsf73w6bspmxgjm0000gn/T/ipykernel_81140/844583093.py:12: DeprecationWarning: Importing clear_output from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML, clear_output


In [5]:
pyrogram.__version__

'2.0.51'

In [6]:
dotenv_path = os.path.join('dot.env')
if os.path.exists(dotenv_path):
    load_dotenv(dotenv_path)

In [7]:
API_ID = os.environ['API_ID']
API_HASH = os.environ['API_HASH']

In [8]:
# выгружаем данные после работы скрипта
# %%bash
# python3 pyro_run.py

In [9]:
# читаем и сохраняем данные в переменную
telegram = pd.read_csv('telegram.csv')

In [10]:
telegram.head(3)

chat  message_id  \
0  {\n    "_": "Chat",\n    "id": -1001137236002,...        4730   
1  {\n    "_": "Chat",\n    "id": -1001137236002,...        4729   
2  {\n    "_": "Chat",\n    "id": -1001137236002,...        4728   

                  date                                               text  \
0  2022-09-09 17:05:10  Product Analyst\nв Safeguard Global — поставщи...   
1  2022-09-09 15:12:08  Бизнес-аналитик/Менеджер проекта\nв Simble — I...   
2  2022-09-09 09:00:11  Lead Data Engineer\nв Cardsmobile — компания р...   

                                            entities  
0  [\n    {\n        "_": "MessageEntity",\n     ...  
1  [\n    {\n        "_": "MessageEntity",\n     ...  
2  [\n    {\n        "_": "MessageEntity",\n     ...

In [11]:
telegram.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8046 entries, 0 to 8045
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   chat        5048 non-null   object
 1   message_id  8046 non-null   int64 
 2   date        8046 non-null   object
 3   text        7512 non-null   object
 4   entities    7345 non-null   object
dtypes: int64(1), object(4)
memory usage: 314.4+ KB


In [12]:
# избавляемся от пропусков
telegram = telegram.dropna().reset_index(drop=True) 

In [13]:
telegram.shape

(4580, 5)

В отдельную колонку сохраним информацию о id чата и его названии.

In [14]:
telegram["chat_id"] = telegram["chat"].apply(lambda x: json.loads(x)['id'])
telegram["title_chat"] = telegram["chat"].apply(lambda x: json.loads(x)['title'])

In [15]:
# В колонке data изменит тип данных
telegram.date = pd.to_datetime(telegram['date'])

In [16]:
telegram.date.sort_values(ascending=False)

0      2022-09-09 17:05:10
1      2022-09-09 15:12:08
847    2022-09-09 14:39:26
2850   2022-09-09 13:46:35
1324   2022-09-09 10:02:44
               ...        
4320   2019-02-18 13:24:38
4321   2019-02-10 12:12:21
4322   2019-02-06 14:14:56
4323   2019-01-31 22:29:30
4324   2019-01-29 20:02:49
Name: date, Length: 4580, dtype: datetime64[ns]

In [17]:
telegram.head(1)

chat  message_id  \
0  {\n    "_": "Chat",\n    "id": -1001137236002,...        4730   

                 date                                               text  \
0 2022-09-09 17:05:10  Product Analyst\nв Safeguard Global — поставщи...   

                                            entities        chat_id  \
0  [\n    {\n        "_": "MessageEntity",\n     ... -1001137236002   

                           title_chat  
0  Job for Analysts & Data Scientists

Кнопка для просмотра выгруженных вакансий.

In [18]:
button = widgets.Button(description="Показать")
output = widgets.Output()

display(button, output)

counter = 0
def on_clicked(b):
    with output:
        try:
            global counter
            clear_output()
            print('№ вакансии: ', counter)
            print(telegram['text'][counter])
            counter += 1
        except Exception as e:
            print("Вакансии закончились")
            
button.on_click(on_clicked)

Button(description='Показать', style=ButtonStyle())

Output()

## Вывод
* Данные о вакансиях парсили из 7 телеграм-каналов
* Всего 4580 строк
* Данные выгруженны за период с 29.01.2019 по 09.09.2022

# Извлечение информации по ссылкам в вакансиях

## Выгрузим ссылки в отдельную колонку датасета - `url`

In [19]:
# def get_links(x, regexp='http\S+'): ###
#     try:
#         for i in json.loads(x['entities']):
#             if i['type'] == 'MessageEntityType.TEXT_LINK':
#                 return i['url']
#             elif i['type'] == 'MessageEntityType.URL':
#                 url = re.findall(regexp, x['text'])[0]
#                 return url
#     except:
#         return None

In [20]:
def get_links(x, regexp='http\S+'): ###
    try:
        for i in json.loads(x['entities']):
            if i['type'] == 'MessageEntityType.TEXT_LINK':
                return i['url']
            elif i['type'] == 'MessageEntityType.URL':
                url = re.findall(regexp, x['text'])[0]
                return url
    except:
        return None

In [21]:
telegram['url'] = telegram.apply(get_links, axis=1)

In [22]:
telegram['url'][0]

'https://www.linkedin.com/jobs/view/3243717640/'

## Найдем наиболее встречающиеся сайты, с которых поступает информация о вакансиях

In [23]:
# функция для поиска hostname url ссылки
def get_hostname(link):
    o = urlparse(link)
    host = o.hostname
    return host

In [24]:
telegram['host'] = telegram['url'].apply(lambda x: get_hostname(x))

Выведем 30 hostname, ссылка на которые встречаются больше всего. Будем вытаскивать информацию из ссылки, если она встречается больше 5 раз и ведет на сайт с вакансиями. 

In [25]:
telegram.groupby('host')['text'].agg('count').sort_values(ascending=False).head(30)

host
geekjob.ru                                   733
ya.cc                                        203
hh.ru                                        131
t.me                                         115
purninja-st.s3.eu-central-1.amazonaws.com     40
is.gd                                         31
otus.ru                                       26
i.ibb.co                                      26
spb.hh.ru                                     24
practicum.yandex.ru                           23
www.notion.so                                 23
telegra.ph                                    14
proglib.io                                    14
clck.ru                                       14
www.youtube.com                               12
ctrl2go.solutions                             12
praktikum.yandex.ru                           12
www.linkedin.com                              10
netology.ru                                    9
yandex.ru                                      9
your.gms.tech  

## Напишем функции для парсинга ссылок

In [26]:
class RequestVacancy:
    USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36'
    
    def get_vacancy(self, url):
        if not url:
            return
        
        try:
            return requests.get(
                url, 
                headers=self._get_headers(),
#                 verify=False,
                timeout=20,
            )
        except ConnectionError:
            print(f'ConnectionError:\t{url}')
        except TimeoutError:
            print(f'TimeoutError:\t{url}')
        except InvalidSchema:
            print(f'InvalidSchema:\t{url}')
        except Exception as e:
            print(f'{type(e)}:\t{url}')
    
    def _get_headers(self) -> dict:
        return {
            'User-Agent': self.USER_AGENT,
        }

# vacancy requester
vc = RequestVacancy()

In [27]:
# def parse_site(url, element='div', extra_info=None):
#     try:
#         data = requests.get(url).text 
#         soup = BeautifulSoup(data)
#         elements = []
#         for extra in extra_info:
#             elements += soup.find_all(element, extra)
#         html = ''
#         for i in elements:
#             html += str(i)
#         return html
#     except Exception as e:
#         return '<p>Не удалось получить элемент</p>'

In [28]:
# HTML(parse_site(telegram['url'][1], element='section', extra_info=[{'class': 'col s12 m12 main'}]))

Функции для парсинга ссылок

In [29]:
from requests.exceptions import InvalidSchema

def default_parser(url):
    title, company, location, grade, salary, description = None, None, None, None, None, None
    return {
        "title": title,
        "comapany": company,
        "location": location,
        "grade": grade,
        "salary": salary,
        "description": description
    }


def _get_domain(url):
    """
    'https://www.linkedin.com/jobs/view/2865172210/' => 'www.linkedin.com'
    """
    return '.'.join(urlparse(url).netloc.split('.')[-2:])


def get_text_from_url(url):
    response = vc.get_vacancy(url)
    if not response:
        return
    if response.status_code > 305:
        print(f'Status code: {response.status_code}')
        return
    return response.text



def apply_info(url):
    text = get_text_from_url(url)
    if not text:
        return
    
    parser = parsers.get(_get_domain(url))
    if not parser:
        parser = parsers['default']
    
    final_text = None
#     final_text = parser(text)
#     return json.dumps(final_text)
    try:
        final_text = parser(text)
    except Exception as e:
        print(f'{type(e)}:\t{e}:\t{url}')
    else:
        return json.dumps(final_text)

In [30]:
# для сайта geekjob
def parse_geekjob(text):
    soup = BeautifulSoup(text) 
    
    title = soup.find('h1').text.strip()
    company = soup.select_one('a#company-web').text.strip()
    location = soup.find('div', {'class':"location"}).text.strip() 
    grade = soup.find('div', {'class': "category"}).text.strip()
    salary = soup.find('span', {'class': "salary"}).text.strip()
    description = soup.find('div', {'class': "description"}).text.strip()
        
    return {
        "title": title,
        "comapany": company,
        "location": location,
        "grade": grade,
        "salary": salary,
        "description": description
    }

In [31]:
# для сайта hh.ru
def parse_hh(text):
    
    soup = BeautifulSoup(text)

    title = soup.find('h1', {'data-qa':"vacancy-title"}).text.strip() #*
    company = soup.find('span', {'class': "vacancy-company-name"}, limit=1).text.strip()
    location = soup.find('p', {'data-qa':"vacancy-view-location"}).text.strip()
    grade = soup.find('p', {'class': "vacancy-description-list-item"}, limit=1).text.strip()
    salary = soup.find('div', {'data-qa': "vacancy-salary"}, limit=1).text.strip() #*
    description = soup.find('div', {'class': "vacancy-section"}).text.strip()
#    title = re.sub(r'[^а-яА-Яa-zA-Z]', ' ', title)
    
    return {
        "title": title,
        "comapany": company,
        "location": location,
        "grade": grade,
        "salary": salary,
        "description": description
    }

In [32]:
# для сайта proglib
def parse_proglib(text):
    
    soup = BeautifulSoup(text)
    
    title = soup.find('h1', {'class':"entity__title mb-2"}).text.strip() #*
    company = soup.find('span', {'itemprop': "hiringOrganization"}).text.strip()
    location = soup.find('p', {'data-qa':"vacancy-view-location"}).text.strip()
    grade = soup.find('span', {'itemprop': "experienceRequirements"}).text.strip()
    salary = soup.find('span', {'itemprop': "baseSalary"}).text.strip() #*
    description = soup.find('div', {'class': "block__content entity__content"}).text.strip()
#    title = re.sub(r'[^а-яА-Яa-zA-Z]', ' ', title)
    
    return {
        "title": title,
        "comapany": company,
        "location": location,
        "grade": grade,
        "salary": salary,
        "description": description
    }

In [33]:
# для сайта linkedn
def parse_linkedn(text):
    
    soup = BeautifulSoup(text)
    
    title = soup.find('h1', {'class':"top-card-layout__title"}).text.strip() #*
    company = soup.find('a', {'class': "topcard__org-name-link"}).text.strip()
    location = soup.find('span', {'class':"topcard__flavor topcard__flavor--bullet"}).text.strip()
    grade = soup.find('h3', {'class': "description__job-criteria-subheader"}).text.strip()
    salary = 'None' #*
    description = soup.find('div', {'class': "show-more-less-html__markup"}).text.strip()
#    title = re.sub(r'[^а-яА-Яa-zA-Z]', ' ', title)
    
    return {
        "title": title,
        "comapany": company,
        "location": location,
        "grade": grade,
        "salary": salary,
        "description": description
    }

In [34]:
# для сайта yandex
def parse_yandex(text):
    
    soup = BeautifulSoup(text)

    title = soup.find('h1', {'class':"lc-styled-text__text"}).text.strip() #*
    company = 'Yandex'
    #     location = soup.find('span', {'class':"Text Text_typography_control-s lc-jobs-tag__label"}).text.strip()
    #     grade = soup.find('div', {'class': "lc-jobs-vacancy__meta-section-items"}).text.strip()

    location_tag = soup.select_one('div.lc-jobs-vacancy__meta-section:nth-child(3)')
    location_items = location_tag.select('span.lc-jobs-tag__label')
    location = ', '.join((t.text for t in location_items))

    grade_items = (
            soup
            .select_one('div.lc-jobs-vacancy__meta-section:nth-child(5)')
            .select('span.lc-jobs-tag__label')
        )
    grade = ', '.join((t.text for t in grade_items))


    salary = 'None' #*
    description = soup.find('div', {'class': "lc-jobs-vacancy__description"}).text.strip()
 
    #title = re.sub(r'[^а-яА-Яa-zA-Z]', ' ', title)
    
    return {
        "title": title,
        "comapany": company,
        "location": location,
        "grade": grade,
        "salary": salary,
        "description": description
    }

In [35]:
parsers = {
    'geekjob.ru': parse_geekjob,
    'ya.cc' : parse_yandex,
    'linkedin.com': parse_linkedn,
    'hh.ry': parse_hh,
    'default': default_parser
}

In [36]:
# parse_yandex(get_text_from_url('https://yandex.ru/jobs/vacancies/%D1%80%D0%B0%D0%B7%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D1%87%D0%B8%D0%BA-%D0%B8%D1%81%D1%81%D0%BB%D0%B5%D0%B4%D0%BE%D0%B2%D0%B0%D1%82%D0%B5%D0%BB%D1%8C-%D1%81%D0%B8%D0%BC%D1%83%D0%BB%D1%8F%D1%86%D0%B8%D0%B8-%D0%B0%D0%B3%D0%B5%D0%BD%D1%82%D0%BE%D0%B2-%D0%B2-%D1%81%D0%B5%D1%80%D0%B2%D0%B8%D1%81-%D0%B1%D0%B5%D1%81%D0%BF%D0%B8%D0%BB%D0%BE%D1%82%D0%BD%D1%8B%D1%85-%D0%B0%D0%B2%D1%82%D0%BE%D0%BC%D0%BE%D0%B1%D0%B8%D0%BB%D0%B5%D0%B9-5773'))

In [37]:
# parse_linkedn(get_text_from_url('https://www.linkedin.com/jobs/collections/recommended/?currentJobId=3253779996'))

In [38]:
from tqdm import tqdm
tqdm.pandas()
#df_users.groupby(['userID', 'requestDate']).progress_apply(feature_rollup)

In [39]:
telegram['info'] = telegram['url'].progress_apply(lambda x: apply_info(x))

  0%|          | 14/4580 [00:36<2:54:15,  2.29s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://www.linkedin.com/jobs/view/3247021542/


  0%|          | 15/4580 [00:37<2:12:09,  1.74s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sfQ


  2%|▏         | 70/4580 [03:55<4:01:04,  3.21s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sdC


  2%|▏         | 88/4580 [05:00<3:32:01,  2.83s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sd7


  2%|▏         | 89/4580 [05:01<2:41:55,  2.16s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sck


  2%|▏         | 90/4580 [05:03<2:40:29,  2.14s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/scc


  2%|▏         | 97/4580 [05:27<3:50:17,  3.08s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sce


  2%|▏         | 98/4580 [05:29<3:29:16,  2.80s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/scd


  2%|▏         | 103/4580 [05:40<2:46:51,  2.24s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sbW


  2%|▏         | 104/4580 [05:41<2:23:31,  1.92s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sbV


  2%|▏         | 105/4580 [05:43<2:17:16,  1.84s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sbU


  2%|▏         | 106/4580 [05:45<2:20:12,  1.88s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sbT


  2%|▏         | 107/4580 [05:47<2:33:01,  2.05s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/saX


  2%|▏         | 108/4580 [05:50<2:46:17,  2.23s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/saY


  2%|▏         | 109/4580 [05:52<2:53:38,  2.33s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sYR


  2%|▏         | 110/4580 [05:55<2:54:55,  2.35s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sYS


  3%|▎         | 118/4580 [06:24<4:03:40,  3.28s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sYQ


  3%|▎         | 119/4580 [06:27<3:41:05,  2.97s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sYP


  3%|▎         | 135/4580 [07:18<2:45:40,  2.24s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sV1


  3%|▎         | 136/4580 [07:21<2:47:25,  2.26s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sV5


  3%|▎         | 137/4580 [07:23<2:43:15,  2.20s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sV4


  3%|▎         | 139/4580 [07:27<2:30:48,  2.04s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sWj


  3%|▎         | 155/4580 [08:16<4:13:40,  3.44s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sSX


  3%|▎         | 157/4580 [08:22<3:52:42,  3.16s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sSY


  3%|▎         | 158/4580 [08:24<3:41:15,  3.00s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sSW


  3%|▎         | 159/4580 [08:27<3:32:41,  2.89s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sSZ


  4%|▎         | 163/4580 [08:40<3:30:08,  2.85s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sST


  4%|▎         | 164/4580 [08:42<3:22:20,  2.75s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sR3


  4%|▎         | 167/4580 [08:48<2:52:15,  2.34s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sR4


  4%|▍         | 185/4580 [09:48<3:28:18,  2.84s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sR2


  4%|▍         | 189/4580 [09:59<2:58:46,  2.44s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sRU


  4%|▍         | 194/4580 [10:12<3:12:51,  2.64s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sR5


  4%|▍         | 195/4580 [10:15<3:11:31,  2.62s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sR6


  4%|▍         | 200/4580 [10:30<4:10:36,  3.43s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMj


  4%|▍         | 202/4580 [10:32<2:26:05,  2.00s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMi


  4%|▍         | 203/4580 [10:34<2:36:20,  2.14s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMo


  4%|▍         | 205/4580 [10:42<3:29:53,  2.88s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMg


  4%|▍         | 206/4580 [10:43<2:53:03,  2.37s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMJ


  5%|▍         | 207/4580 [10:46<3:09:04,  2.59s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMe


  5%|▍         | 209/4580 [10:52<3:04:50,  2.54s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMh


  5%|▍         | 215/4580 [11:07<2:11:48,  1.81s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sMD


  5%|▍         | 218/4580 [11:17<3:04:54,  2.54s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sQG


  5%|▍         | 226/4580 [11:53<4:22:40,  3.62s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sKx


  5%|▍         | 227/4580 [11:56<3:57:07,  3.27s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sJd


  5%|▍         | 228/4580 [11:56<2:58:45,  2.46s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sL3


  5%|▌         | 230/4580 [12:02<3:17:29,  2.72s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sK4


  5%|▌         | 231/4580 [12:05<3:14:03,  2.68s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sJY


  5%|▌         | 237/4580 [12:33<4:28:41,  3.71s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sJZ


  5%|▌         | 238/4580 [12:35<3:51:38,  3.20s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sJj


  5%|▌         | 243/4580 [12:49<3:39:11,  3.03s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sJe


  5%|▌         | 247/4580 [13:03<3:25:34,  2.85s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sJb


  5%|▌         | 248/4580 [13:05<3:17:34,  2.74s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sJW


  6%|▌         | 255/4580 [13:29<3:37:45,  3.02s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sIo


  6%|▌         | 256/4580 [13:31<3:24:19,  2.84s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sIx


  6%|▌         | 257/4580 [13:33<3:10:52,  2.65s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sIt


  6%|▌         | 258/4580 [13:36<3:07:16,  2.60s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sIw


  6%|▌         | 275/4580 [14:44<4:17:57,  3.60s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sIK


  6%|▌         | 276/4580 [14:46<3:39:03,  3.05s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sHw


  6%|▌         | 277/4580 [14:50<4:12:33,  3.52s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sHv


  6%|▌         | 279/4580 [14:55<3:17:22,  2.75s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sID


  6%|▌         | 281/4580 [14:58<2:44:58,  2.30s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sHu


  6%|▌         | 282/4580 [15:01<2:50:17,  2.38s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sHt


  6%|▌         | 284/4580 [15:09<3:47:24,  3.18s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sHK


  6%|▋         | 292/4580 [15:44<4:26:46,  3.73s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sHI


  7%|▋         | 301/4580 [16:10<3:02:15,  2.56s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://blog.geekjob.ru/digest-analyst-may-22/


  7%|▋         | 312/4580 [16:53<5:50:31,  4.93s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sDC


  7%|▋         | 313/4580 [16:56<5:02:43,  4.26s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sDc


  7%|▋         | 317/4580 [17:28<7:47:47,  6.58s/it] 

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sDB


  8%|▊         | 365/4580 [20:05<3:06:32,  2.66s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sB2


  8%|▊         | 368/4580 [20:13<2:55:30,  2.50s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sB0


  8%|▊         | 384/4580 [20:52<1:39:57,  1.43s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/sB5


  9%|▉         | 428/4580 [23:23<3:10:18,  2.75s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s8O


  9%|▉         | 431/4580 [23:29<2:24:45,  2.09s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s8M


  9%|▉         | 433/4580 [23:35<2:42:42,  2.35s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s8L


  9%|▉         | 435/4580 [23:40<2:52:03,  2.49s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s8J


 10%|▉         | 437/4580 [23:46<2:59:27,  2.60s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s8I


 10%|▉         | 442/4580 [24:03<3:29:01,  3.03s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s7B


 10%|▉         | 449/4580 [24:19<2:23:00,  2.08s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s79


 10%|▉         | 451/4580 [24:25<2:45:15,  2.40s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s7C


 12%|█▏        | 570/4580 [30:14<2:37:58,  2.36s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/hXUr


 15%|█▌        | 702/4580 [37:27<3:51:54,  3.59s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/hX7Z


 17%|█▋        | 760/4580 [40:53<3:03:35,  2.88s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s3E


 18%|█▊        | 847/4580 [45:53<2:45:29,  2.66s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://geekjob.ru/s1


 22%|██▏       | 999/4580 [46:32<08:21,  7.13it/s]  

<class 'requests.exceptions.ConnectionError'>:	https://www.in-aim.ru)


 23%|██▎       | 1059/4580 [46:38<05:16, 11.11it/s]

<class 'requests.exceptions.ConnectionError'>:	https://selectel.ru)
<class 'requests.exceptions.ConnectionError'>:	https://tkmistral.ru)


 29%|██▉       | 1339/4580 [48:02<1:03:12,  1.17s/it]

InvalidSchema:	tg://search_hashtag?hashtag=%D0%B2%D0%B0%D0%BA%D0%B0%D0%BD%D1%81%D0%B8%D1%8F


 31%|███       | 1402/4580 [48:21<13:36,  3.89it/s]  

<class 'requests.exceptions.ConnectionError'>:	https://company.myseldon.com)


 32%|███▏      | 1448/4580 [48:49<24:46,  2.11it/s]  

InvalidSchema:	tg://search_hashtag?hashtag=fulltime


 32%|███▏      | 1461/4580 [48:52<10:57,  4.75it/s]

InvalidSchema:	tg://search_hashtag?hashtag=vacancy


 32%|███▏      | 1463/4580 [48:55<30:44,  1.69it/s]

<class 'requests.exceptions.ConnectionError'>:	https://anyquery.diginetica.com).


 35%|███▍      | 1583/4580 [49:45<29:34,  1.69it/s]  

<class 'requests.exceptions.ConnectionError'>:	http://aclyon.ru)


 35%|███▍      | 1586/4580 [49:47<29:45,  1.68it/s]

<class 'requests.exceptions.ConnectionError'>:	https://onesoil.ai,


 35%|███▌      | 1611/4580 [49:57<20:55,  2.36it/s]

<class 'requests.exceptions.ConnectionError'>:	https://neuton.ai)


 37%|███▋      | 1673/4580 [50:07<06:40,  7.26it/s]

<class 'requests.exceptions.ConnectionError'>:	https://www.ngrow.ai)


 38%|███▊      | 1740/4580 [50:29<08:40,  5.45it/s]

<class 'requests.exceptions.ConnectionError'>:	https://mipt.x5.ru)


 39%|███▉      | 1785/4580 [50:56<13:56,  3.34it/s]

<class 'requests.exceptions.SSLError'>:	http://sber.me/?p=W1112


 40%|████      | 1847/4580 [51:09<07:36,  5.98it/s]

<class 'requests.exceptions.ConnectionError'>:	https://cloud.lenaneva.ru/apps/files_sharing/publicpreview/LQZMxywLPZiecis?x=2714&y=1074&a=true&file=multiarmed_0402.jpg&scalingup=0


 40%|████      | 1849/4580 [51:10<07:39,  5.95it/s]

<class 'requests.exceptions.ConnectionError'>:	https://cloud.lenaneva.ru/apps/files_sharing/publicpreview/ifaWwiZ29yTXr7A?x=2714&y=1074&a=true&file=otus_spark0102.png&scalingup=0
<class 'requests.exceptions.ConnectionError'>:	https://cloud.lenaneva.ru/apps/files_sharing/publicpreview/n226PyGqSpXdmWa?x=2714&y=1074&a=true&file=mla_2701.jpg&scalingup=0


 41%|████      | 1856/4580 [51:10<05:55,  7.67it/s]

<class 'requests.exceptions.ConnectionError'>:	https://cloud.lenaneva.ru/apps/files_sharing/publicpreview/CCXDBdQKkt2fX4C?x=2714&y=1074&a=true&file=otus_machine_learning_advanced_2201.jpg&scalingup=0


 41%|████      | 1883/4580 [51:36<42:42,  1.05it/s]

<class 'requests.exceptions.ConnectionError'>:	https://identix.one/


 47%|████▋     | 2155/4580 [52:35<13:04,  3.09it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/WTxgYI1P3XJh2p


 47%|████▋     | 2160/4580 [52:36<11:55,  3.38it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/P69wtulR3TPYsG


 47%|████▋     | 2162/4580 [52:38<17:11,  2.34it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/hWuzIQLX3SjxEi


 48%|████▊     | 2187/4580 [52:44<07:38,  5.22it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/Qk5IO5Bj3QAQpx


 48%|████▊     | 2208/4580 [52:47<05:08,  7.70it/s]

InvalidSchema:	tg://search_hashtag?hashtag=fulltime


 48%|████▊     | 2213/4580 [52:49<08:34,  4.60it/s]

InvalidSchema:	tg://search_hashtag?hashtag=vacancy
<class 'requests.exceptions.ConnectionError'>:	https://anyquery.diginetica.com).


 49%|████▉     | 2234/4580 [52:54<14:08,  2.77it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/jCWwyO-A3CZsz6


 49%|████▉     | 2235/4580 [52:55<17:39,  2.21it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/-VPOkqRw3CRuSx


 49%|████▉     | 2249/4580 [53:06<18:58,  2.05it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/pkzqOvqD3BJtgS


 49%|████▉     | 2250/4580 [53:06<21:31,  1.80it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/G1oxzzKY3BJkD4


 49%|████▉     | 2257/4580 [53:08<12:52,  3.01it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/xVbrOPIN35YV9H


 49%|████▉     | 2258/4580 [53:10<22:28,  1.72it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/BdEgyjMI3AspR9


 49%|████▉     | 2259/4580 [53:12<33:21,  1.16it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/VHtMhr1R3ApBA9


 49%|████▉     | 2264/4580 [53:14<20:52,  1.85it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/hbHR6DZI39VJC3


 50%|████▉     | 2278/4580 [53:20<18:28,  2.08it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/xVbrOPIN35YV9H


 50%|████▉     | 2279/4580 [53:22<28:30,  1.35it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/aY26o2Fj34MyVX


 50%|████▉     | 2287/4580 [53:26<15:39,  2.44it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/YDYwQbCv32qJ4N


 50%|████▉     | 2289/4580 [53:27<16:32,  2.31it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/3V0R0lv932q65q


 50%|█████     | 2291/4580 [53:29<23:24,  1.63it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/vZPOUDNMrkAE2


 50%|█████     | 2295/4580 [53:30<15:42,  2.42it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/Fj5RkHeP32ccQh


 50%|█████     | 2298/4580 [53:31<15:01,  2.53it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/UEM8fA7W32bqCM


 50%|█████     | 2301/4580 [53:33<19:21,  1.96it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/zLaJzQima6Qio


 50%|█████     | 2305/4580 [53:36<26:56,  1.41it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/qrko6ej-yGxKp


 50%|█████     | 2308/4580 [53:38<24:44,  1.53it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/7taaMNk6sp5qs


 51%|█████     | 2316/4580 [53:40<15:55,  2.37it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/qfYq2wByx58tu


 51%|█████     | 2318/4580 [53:42<21:26,  1.76it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/oNBWC6rOx58LT


 51%|█████     | 2322/4580 [53:45<23:31,  1.60it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/nsdUtD6uvmoqH


 51%|█████     | 2323/4580 [53:46<29:50,  1.26it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/YwvINYmFvmpe5


 51%|█████     | 2328/4580 [53:49<28:58,  1.30it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/48fwFBQUuTBCs


 51%|█████     | 2333/4580 [53:52<25:36,  1.46it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/zBos9TyYtcxUU


 51%|█████     | 2334/4580 [53:54<28:54,  1.30it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/7taaMNk6sp5qs


 51%|█████     | 2335/4580 [53:55<32:25,  1.15it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/wOSADVH9sp56x


 51%|█████▏    | 2350/4580 [54:06<32:29,  1.14it/s]

<class 'requests.exceptions.ConnectionError'>:	http://aclyon.ru)


 51%|█████▏    | 2353/4580 [54:09<36:50,  1.01it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/vZPOUDNMrkAE2
<class 'requests.exceptions.ConnectionError'>:	https://onesoil.ai,


 51%|█████▏    | 2356/4580 [54:12<39:23,  1.06s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/MSxB3Ahvfh6Cp


 52%|█████▏    | 2359/4580 [54:13<26:51,  1.38it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/35-R3mAFpjtGo


 52%|█████▏    | 2367/4580 [54:19<30:04,  1.23it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/0y8cOsqnrVDdV


 52%|█████▏    | 2374/4580 [54:24<22:00,  1.67it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/fDQnzcaKoVx4m


 52%|█████▏    | 2378/4580 [54:26<19:27,  1.89it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/R_fFMzCDha9um


 52%|█████▏    | 2380/4580 [54:27<19:19,  1.90it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/nVaqgJe-oQtj5


 52%|█████▏    | 2381/4580 [54:28<23:12,  1.58it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/SWdrId05oQzEg


 52%|█████▏    | 2389/4580 [54:33<27:24,  1.33it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/zFpGTZ6JoQxBB


 52%|█████▏    | 2395/4580 [54:38<24:40,  1.48it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/zLaJzQima6Qio


 52%|█████▏    | 2397/4580 [54:39<24:00,  1.51it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/Q9reaseWSTdN4


 52%|█████▏    | 2398/4580 [54:40<29:20,  1.24it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/k8vRAWJ6ZwiyQ


 52%|█████▏    | 2402/4580 [54:43<29:16,  1.24it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/Dpckv24znEjZo


 52%|█████▏    | 2403/4580 [54:44<30:52,  1.17it/s]

<class 'requests.exceptions.ConnectionError'>:	https://neuton.ai)


 53%|█████▎    | 2413/4580 [54:48<16:30,  2.19it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/IFwtZ5RDn4Gif


 53%|█████▎    | 2421/4580 [54:53<22:05,  1.63it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/QGmp71kwm7wtQ


 53%|█████▎    | 2426/4580 [54:56<22:02,  1.63it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/6KlIJy05kYJQX


 53%|█████▎    | 2428/4580 [54:58<24:54,  1.44it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/oWeiS6cZkPZCZ


 53%|█████▎    | 2437/4580 [55:03<20:02,  1.78it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/k8vRAWJ6ZwiyQ


 53%|█████▎    | 2439/4580 [55:05<23:10,  1.54it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/c6LrVw9IaCmuf


 53%|█████▎    | 2445/4580 [55:08<17:33,  2.03it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/4qrhyI9QbKubX


 53%|█████▎    | 2448/4580 [55:09<15:44,  2.26it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/NTJRBu9giAsje


 53%|█████▎    | 2449/4580 [55:11<20:35,  1.73it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/R_fFMzCDha9um


 54%|█████▎    | 2459/4580 [55:14<12:19,  2.87it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/MfVy6RH3eTBx8


 54%|█████▍    | 2469/4580 [55:19<20:50,  1.69it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/dJKea3YGdJDox


 54%|█████▍    | 2475/4580 [55:23<20:19,  1.73it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/ZQBE-1-HdHRXR


 54%|█████▍    | 2478/4580 [55:24<17:05,  2.05it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/e8fD270Qct6pj


 54%|█████▍    | 2486/4580 [55:26<11:22,  3.07it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/77bLH7Fmct8Xe


 54%|█████▍    | 2489/4580 [55:28<14:58,  2.33it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/MSxB3Ahvfh6Cp


 54%|█████▍    | 2493/4580 [55:30<16:36,  2.09it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/tB8z4t8mccWMi


 55%|█████▍    | 2502/4580 [55:33<14:17,  2.42it/s]

<class 'requests.exceptions.ConnectionError'>:	https://www.ngrow.ai)


 55%|█████▍    | 2512/4580 [55:37<13:38,  2.53it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/nKzA4zyrbKvzf


 55%|█████▍    | 2513/4580 [55:38<19:18,  1.78it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/4qrhyI9QbKubX


 55%|█████▍    | 2515/4580 [55:40<26:39,  1.29it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/KEmYpLhhau7b6


 55%|█████▌    | 2522/4580 [55:43<19:08,  1.79it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/SRIMGvIEakMy5


 55%|█████▌    | 2528/4580 [55:46<16:14,  2.11it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/c6LrVw9IaCmuf


 55%|█████▌    | 2534/4580 [55:50<20:51,  1.64it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/qeL316VfaAG2o


 55%|█████▌    | 2536/4580 [55:51<22:14,  1.53it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/v69pSnEjaA9n4


 55%|█████▌    | 2539/4580 [55:53<20:30,  1.66it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/zLaJzQima6Qio


 56%|█████▌    | 2573/4580 [56:18<33:16,  1.01it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/FH4aswLVXqnbF


 56%|█████▋    | 2582/4580 [56:23<25:45,  1.29it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/gKDkZax6XinJj


 57%|█████▋    | 2589/4580 [56:27<24:04,  1.38it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/5F5SbzCqYRv89


 57%|█████▋    | 2592/4580 [56:35<1:00:20,  1.82s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/5F5SbzCqYRv89


 57%|█████▋    | 2614/4580 [56:50<26:24,  1.24it/s]  

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/ql59BWR2WqAzF


 57%|█████▋    | 2622/4580 [56:59<40:27,  1.24s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/VscgFMiyXiwAC


 57%|█████▋    | 2627/4580 [57:03<34:51,  1.07s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/l2rF-pEvW8JPw


 58%|█████▊    | 2641/4580 [57:06<12:12,  2.65it/s]

<class 'requests.exceptions.SSLError'>:	https://www.bigdataschool.ru/about/jobs-big-data


 58%|█████▊    | 2657/4580 [57:31<44:28,  1.39s/it]  

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/gk6wZipnRap2B


 58%|█████▊    | 2677/4580 [57:38<16:36,  1.91it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/n0es6UtnUJRr7


 58%|█████▊    | 2678/4580 [57:39<21:05,  1.50it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/THW7fcK6UJPQM


 59%|█████▊    | 2680/4580 [57:41<23:01,  1.37it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/_bnyfoe-T6f6A


 59%|█████▊    | 2683/4580 [57:44<29:15,  1.08it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/ZAgm_1NMUDyh9


 59%|█████▊    | 2685/4580 [57:47<34:27,  1.09s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/3Yw06v_lUDefN


 59%|█████▊    | 2686/4580 [57:50<48:48,  1.55s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/89qWZ49UU3N45


 59%|█████▊    | 2687/4580 [57:51<47:59,  1.52s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/u4nSUvG3TsSos


 59%|█████▊    | 2689/4580 [57:53<35:30,  1.13s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/N8uX-GaCTVHGq


 59%|█████▊    | 2690/4580 [57:54<37:49,  1.20s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/Nn5ibKEqTKfVx


 59%|█████▉    | 2691/4580 [57:55<38:54,  1.24s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/nJq4CAFKTURHf


 59%|█████▉    | 2693/4580 [57:58<38:45,  1.23s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/eL60gff3TJLCy


 59%|█████▉    | 2700/4580 [58:04<30:14,  1.04it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/hk_fpognSxHN5


 59%|█████▉    | 2701/4580 [58:07<42:22,  1.35s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/RMAyiiBYSxFw4


 59%|█████▉    | 2703/4580 [58:10<49:19,  1.58s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/l5GN5YQDSxBaK


 59%|█████▉    | 2706/4580 [58:14<42:54,  1.37s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/QrfmylwRSK7bf


 59%|█████▉    | 2707/4580 [58:16<44:54,  1.44s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/-uGaPGfNRaAsA


 59%|█████▉    | 2711/4580 [58:18<30:07,  1.03it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/NcBwJlOuQzUaY


 59%|█████▉    | 2716/4580 [58:23<27:59,  1.11it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/napOo9cySEX4F


 59%|█████▉    | 2717/4580 [58:25<35:22,  1.14s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'select':	https://ya.cc/t/0jA4kKkeRvCXY


 59%|█████▉    | 2721/4580 [58:30<34:01,  1.10s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/vLdoDpK2Re4VJ


 59%|█████▉    | 2725/4580 [58:31<17:06,  1.81it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/-uGaPGfNRaAsA


 60%|█████▉    | 2728/4580 [58:32<13:07,  2.35it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/od_SPaNBQzVc3


 60%|█████▉    | 2730/4580 [58:34<19:18,  1.60it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/GSO8H_pOQzRY5


 60%|█████▉    | 2731/4580 [58:35<21:43,  1.42it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/gdCkcmnCQrMTE


 60%|█████▉    | 2745/4580 [58:40<12:28,  2.45it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/Snq4DW3DQP68o


 60%|█████▉    | 2747/4580 [58:40<11:25,  2.68it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/gNMEZhJjPwXGK


 60%|██████    | 2751/4580 [58:41<08:03,  3.79it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/nDXrm2kgPYuJc


 60%|██████    | 2753/4580 [58:41<07:11,  4.23it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/yXhrJvm3PYgzq


 60%|██████    | 2755/4580 [58:43<14:10,  2.15it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/-VPuXYAsPSRa4


 60%|██████    | 2759/4580 [58:44<09:24,  3.23it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/BPk-cTMENg5S7


 60%|██████    | 2764/4580 [58:44<05:14,  5.78it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/-gWtNbaJQPcau
<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/5XBPOa0jPVqNq


 60%|██████    | 2766/4580 [58:45<04:44,  6.38it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/xWOo3qGTPNp7L


 60%|██████    | 2769/4580 [58:46<06:32,  4.62it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/0bpDCRuIPJUnb


 61%|██████    | 2775/4580 [58:46<03:35,  8.39it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/9Pj2vtBTPBh4q


 61%|██████    | 2777/4580 [58:46<03:33,  8.46it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/PgArzUzYP689E


 61%|██████    | 2781/4580 [58:46<03:37,  8.26it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/RblOVlVSNkicv


 61%|██████    | 2782/4580 [58:47<04:18,  6.96it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/UFQIOHrrNkcP3


 61%|██████    | 2784/4580 [58:51<21:58,  1.36it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/_0zSfZ_3Ndk7t


 61%|██████    | 2789/4580 [58:52<10:10,  2.94it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/83RcNauvNkdyJ


 61%|██████    | 2793/4580 [58:57<25:08,  1.18it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/rOQ3geeLGDAih


 61%|██████    | 2795/4580 [58:57<17:15,  1.72it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/3GNm2V0mNd8nT


 61%|██████    | 2796/4580 [58:57<14:45,  2.02it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	http://ya.cc/t/osnJ5P4dJSPk3


 61%|██████    | 2798/4580 [58:58<10:01,  2.96it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/URFrchTENcdvm


 61%|██████    | 2801/4580 [59:06<40:35,  1.37s/it]  

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/BP4gjYWHMtfff


 61%|██████    | 2804/4580 [59:07<22:34,  1.31it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/A4Qvi2cCMtfsq


 61%|██████▏   | 2806/4580 [59:11<33:07,  1.12s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/eEZ5aiuuMte8j


 61%|██████▏   | 2808/4580 [59:14<35:45,  1.21s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/H2yeS3NEMjxH7


 61%|██████▏   | 2811/4580 [59:34<2:37:31,  5.34s/it]

<class 'requests.exceptions.ConnectTimeout'>:	https://www.mos.ru/news/item/83314073/


 61%|██████▏   | 2812/4580 [59:35<1:59:59,  4.07s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/gfcSCUc5Hem8J


 61%|██████▏   | 2813/4580 [59:35<1:30:19,  3.07s/it]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/HnH6pGstMGmTU


 62%|██████▏   | 2819/4580 [59:36<24:26,  1.20it/s]  

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	http://ya.cc/t/osnJ5P4dJSPk3


 62%|██████▏   | 2820/4580 [59:37<24:14,  1.21it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	http://ya.cc/t/qK9JF3WjMCqGH


 62%|██████▏   | 2828/4580 [59:39<08:40,  3.36it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	http://ya.cc/t/AHIeBhY5LPmMu


 62%|██████▏   | 2832/4580 [59:40<10:29,  2.78it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	http://ya.cc/t/JXpImQTqGPF8W


 62%|██████▏   | 2836/4580 [59:41<07:42,  3.77it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	http://ya.cc/t/DgFWY-BWKujZg


 62%|██████▏   | 2839/4580 [59:42<05:27,  5.31it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/ndhkjow9KnB3T


 62%|██████▏   | 2842/4580 [59:42<04:57,  5.85it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/Sq_onopcKi9VZ


 62%|██████▏   | 2849/4580 [59:44<06:23,  4.52it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/mwU0EpV8KKgFm


 63%|██████▎   | 2905/4580 [59:59<11:22,  2.45it/s]

Status code: 999


 65%|██████▍   | 2959/4580 [1:00:28<36:13,  1.34s/it]

<class 'requests.exceptions.SSLError'>:	https://developers.sber.ru/portal/events/onedayoffer-july


 65%|██████▌   | 2993/4580 [1:00:48<10:02,  2.63it/s]

<class 'requests.exceptions.ConnectionError'>:	https://marketolog.mts.ru)


 66%|██████▌   | 3003/4580 [1:00:55<16:26,  1.60it/s]

<class 'requests.exceptions.ConnectionError'>:	https://rivegauche.ru)


 66%|██████▌   | 3030/4580 [1:01:03<13:11,  1.96it/s]

<class 'requests.exceptions.ConnectionError'>:	https://leroymerlin.ru).


 66%|██████▌   | 3034/4580 [1:01:05<11:15,  2.29it/s]

<class 'requests.exceptions.ConnectionError'>:	https://leroymerlin.ru)


 66%|██████▋   | 3045/4580 [1:01:12<14:22,  1.78it/s]

<class 'requests.exceptions.SSLError'>:	https://is-mmk.ru/)


 67%|██████▋   | 3071/4580 [1:01:18<06:26,  3.90it/s]

<class 'requests.exceptions.ConnectionError'>:	https://leroymerlin.ru)


 67%|██████▋   | 3074/4580 [1:01:19<06:21,  3.95it/s]

<class 'requests.exceptions.ConnectionError'>:	https://leroymerlin.ru)


 67%|██████▋   | 3090/4580 [1:01:24<07:43,  3.22it/s]

<class 'requests.exceptions.ConnectionError'>:	http://nsc.media)


 68%|██████▊   | 3120/4580 [1:01:28<02:36,  9.30it/s]

<class 'requests.exceptions.ConnectionError'>:	https://leroymerlin.ru)


 69%|██████▉   | 3173/4580 [1:01:39<03:14,  7.24it/s]

<class 'requests.exceptions.ConnectionError'>:	https://leroymerlin.ru)


 70%|██████▉   | 3190/4580 [1:01:42<04:09,  5.57it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://ya.cc/t/cikNOPtP3E4QeD


 74%|███████▍  | 3402/4580 [1:02:41<04:07,  4.77it/s]

<class 'requests.exceptions.ConnectionError'>:	https://www.asna.ru)


 76%|███████▌  | 3471/4580 [1:02:57<04:56,  3.74it/s]

<class 'requests.exceptions.SSLError'>:	https://www.tadviser.ru/index.php/%D0%9F%D1%80%D0%BE%D0%B4%D1%83%D0%BA%D1%82:SAP_Data_Intelligence


 76%|███████▌  | 3484/4580 [1:02:58<02:18,  7.90it/s]

Status code: 999


 76%|███████▌  | 3486/4580 [1:02:59<02:51,  6.39it/s]

Status code: 999


 78%|███████▊  | 3578/4580 [1:03:12<01:56,  8.63it/s]

<class 'requests.exceptions.ConnectionError'>:	https://ii-s-nulya.ru/
<class 'requests.exceptions.ConnectionError'>:	https://www.asna.ru)


 78%|███████▊  | 3589/4580 [1:03:14<02:09,  7.68it/s]

<class 'requests.exceptions.ConnectionError'>:	https://www.ngrow.ai)


 79%|███████▉  | 3616/4580 [1:03:22<07:20,  2.19it/s]

<class 'requests.exceptions.ConnectionError'>:	https://www.asna.ru)


 80%|███████▉  | 3645/4580 [1:03:27<02:34,  6.04it/s]

<class 'requests.exceptions.MissingSchema'>:	http,


 80%|███████▉  | 3652/4580 [1:03:27<02:03,  7.51it/s]

Status code: 999


 82%|████████▏ | 3746/4580 [1:03:42<02:56,  4.73it/s]

<class 'requests.exceptions.ConnectionError'>:	https://softwarecountry.ru/


 82%|████████▏ | 3776/4580 [1:03:51<02:40,  5.01it/s]

<class 'requests.exceptions.SSLError'>:	https://www.finstar.com/).


 85%|████████▌ | 3898/4580 [1:04:16<03:25,  3.32it/s]

<class 'requests.exceptions.ConnectionError'>:	https://pik.pro/


 87%|████████▋ | 4005/4580 [1:04:41<02:03,  4.66it/s]

Status code: 999


 88%|████████▊ | 4013/4580 [1:04:47<04:55,  1.92it/s]

<class 'requests.exceptions.ConnectionError'>:	https://cdto.work).


 88%|████████▊ | 4023/4580 [1:04:50<03:17,  2.83it/s]

<class 'requests.exceptions.SSLError'>:	https://www.uralchem.ru/


 89%|████████▊ | 4064/4580 [1:05:07<01:53,  4.55it/s]

<class 'AttributeError'>:	'NoneType' object has no attribute 'text':	https://hr.geekjob.ru/vacancy/edit?id=5ee24b69e90a14514b530df2


 89%|████████▉ | 4070/4580 [1:05:08<01:28,  5.79it/s]

<class 'requests.exceptions.SSLError'>:	https://my.sbertalents.ru/#/job-requisition/777250


 92%|█████████▏| 4213/4580 [1:06:02<07:19,  1.20s/it]

<class 'requests.exceptions.ConnectTimeout'>:	https://www.osahp.com/ru/


 93%|█████████▎| 4259/4580 [1:06:15<01:36,  3.34it/s]

<class 'requests.exceptions.ConnectionError'>:	https://whisk.com)


 97%|█████████▋| 4455/4580 [1:07:38<02:27,  1.18s/it]

<class 'requests.exceptions.ConnectTimeout'>:	https://sizdev.ru/


 97%|█████████▋| 4464/4580 [1:07:40<01:33,  1.24it/s]

<class 'requests.exceptions.ConnectionError'>:	https://strategy-it.com)


 98%|█████████▊| 4477/4580 [1:07:46<01:08,  1.51it/s]

<class 'requests.exceptions.ConnectionError'>:	https://onesoil.ai,


 99%|█████████▊| 4514/4580 [1:08:06<00:31,  2.09it/s]

<class 'requests.exceptions.ConnectionError'>:	https://www.ngrow.ai)


 99%|█████████▉| 4540/4580 [1:08:13<00:12,  3.27it/s]

<class 'requests.exceptions.ConnectionError'>:	https://anyquery.diginetica.com)


100%|██████████| 4580/4580 [1:08:31<00:00,  1.11it/s]


In [64]:
telegram['info'][60]

'{"title": "\\u0421\\u0438\\u0441\\u0442\\u0435\\u043c\\u043d\\u044b\\u0439 \\u0430\\u043d\\u0430\\u043b\\u0438\\u0442\\u0438\\u043a 1\\u0421 ERP", "comapany": "icl-services.com", "location": "\\u041a\\u0430\\u0437\\u0430\\u043d\\u044c, \\u0420\\u043e\\u0441\\u0441\\u0438\\u044f", "grade": "\\u041c\\u0438\\u0434\\u0434\\u043b \\u2022 \\u0421\\u0435\\u043d\\u044c\\u043e\\u0440 \\u2022 \\u041a\\u043e\\u043d\\u0441\\u0443\\u043b\\u044c\\u0442\\u0430\\u043d\\u0442", "salary": "", "description": "\\u041e\\u043f\\u0438\\u0441\\u0430\\u043d\\u0438\\u0435 \\u0432\\u0430\\u043a\\u0430\\u043d\\u0441\\u0438\\u0438\\u0414\\u043b\\u044f \\u0431\\u043b\\u043e\\u043a\\u0430 \\"\\u0411\\u044e\\u0434\\u0436\\u0435\\u0442\\u0438\\u0440\\u043e\\u0432\\u0430\\u043d\\u0438\\u0435\\" welcome-\\u0431\\u043e\\u043d\\u0443\\u0441 \\u0434\\u043e 100 000 gross. \\u0418 \\u0442\\u0430\\u043a\\u0436\\u0435 \\u0437\\u0430 \\u0443\\u0441\\u043f\\u0435\\u0448\\u043d\\u0443\\u044e \\u0440\\u0435\\u043a\\u043e\\u043c\\

In [66]:
telegram.to_csv('telegram_plus_link.csv')

**Очистка текста**

**Избавимся от смайликов и не нужных символов в вакансии**

С помощью регулярных выражений очистим описание вакансий, оставив только английские, русские буквы, цифры.

In [40]:
def clin_rex(text):
    text = re.sub(r'[^а-яА-Яa-zA-Z0-9]', ' ', text)
    text = re.sub(r'[^\w\d\s]', ' ', text)
    return text

In [41]:
telegram['text'] = telegram['text'].apply(lambda x: clin_rex(x))

In [42]:
telegram['text'][0]

'Product Analyst в Safeguard Global   поставщик аутсорсинговых услуг по расчету заработной платы и управлению персоналом  Удаленная работа  Описание вакансии'

In [43]:
LABEL_1 = vac_name
LABEL_2 = type_of_work
LABEL_3 = company
LABEL_4 = stack


TRAIN_DATA = [
    (
        "Product Analyst в Safeguard Global поставщик аутсорсинговых услуг по расчету заработной платы и управлению персоналом Удаленная работа",
        {"entities": [(45, 54, LABEL)]},
    ),
    ("Анализ тестовых данных и данных в системах", {"entities": []}),
    (
        "Можно удаленно",
        {"entities": [(6, 14, LABEL)]},
    ),
    ("Удаленный формат работы", {"entities": [(0, 9, LABEL)]}),
    
]

NameError: name 'vac_name' is not defined